website_sale_private_product_page/models/product_template.py

```diff
@@ -0,0 +1,73 @@
+# SPDX-FileCopyrightText: 2022 Coop IT Easy SC
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+
+import secrets
+import string
+
+from odoo import api, fields, models
+
+_LETTERS_AND_DIGITS = string.ascii_letters + string.digits
+
+
+class ProductTemplate(models.Model):
+    _inherit = "product.template"
+
+    private_url_token = fields.Char(
+        string="Private URL Token",
+        help="""
+            A randomly generated token that can be used by third parties to
+            access the product's e-commerce page even when the product is not
+            published. Make sure to keep the token secret from those whom you do
+            not want to see the product page.""",
+        readonly=True,
+        copy=False,
+    )
+    private_url = fields.Char(
+        string="Private URL",
+        readonly=True,
+        compute="_compute_private_url",
+        store=True,
+    )
+
+    _sql_constraints = [
+        (
+            "private_url_token_unique",
+            "UNIQUE (private_url_token)",
+            "The private URL token must be unique",
+        )
+    ]
+
+    @api.depends("private_url_token")
+    def _compute_private_url(self):
+        for template in self:
+            if not template.private_url_token:
+                template.private_url = False
+            else:
+                template.private_url = "/shop/private_product/{}".format(
+                    template.private_url_token
+                )
+
+    def action_generate_private_url_token(self):
+        count = len(self)
+        while True:
+            tokens = {self._generate_private_url_token() for _ in range(count)}
+            # Check for a once-in-52.036.560.683.837.093.888 event: the same
+            # token being generated twice.
```

Comment:
useful comment. note that in python, the `.` character is the decimal point, not a thousands separator. i would apply this to comments as well. in python 3, `_` can be used as thousands separators, or you can use the u.s. `,` for comments.